File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/r3f-require-shadows-enabled.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getAuthoritativeJsxAttribute } from "../../utils/get-authoritative-jsx-attribute.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { UNSUPPORTED_SHADOW_LIGHT_CONSTRUCTOR_NAMES } from "./constants.js";
import { hasR3fRuntimeImport } from "./utils/has-r3f-runtime-import.js";
import { isR3fCanvas } from "./utils/is-r3f-canvas.js";
import { isR3fHostIntrinsic } from "./utils/is-r3f-host-intrinsic.js";
import { readStaticJsxBooleanAttribute } from "./utils/read-static-jsx-boolean-attribute.js";

const SHADOW_OBJECT_PROPERTY_NAMES: ReadonlySet<string> = new Set(["castShadow", "receiveShadow"]);
const R3F_UNSUPPORTED_SHADOW_LIGHT_NAMES: ReadonlySet<string> = new Set(
  [...UNSUPPORTED_SHADOW_LIGHT_CONSTRUCTOR_NAMES].map(
    (constructorName) => `${constructorName[0]?.toLowerCase()}${constructorName.slice(1)}`,
  ),
);

export const r3fRequireShadowsEnabled = defineRule({
  id: "r3f-require-shadows-enabled",
  title: "R3F shadow caster without Canvas shadow maps",
  category: "Correctness",
  tags: ["react-jsx-only"],
  severity: "warn",
  recommendation:
    "Enable the Canvas shadows prop when descendants cast or receive real-time shadows",
  create: (context: RuleContext) => {
    const shadowUsersByCanvas = new Map<
      EsTreeNodeOfType<"JSXOpeningElement">,
      EsTreeNodeOfType<"JSXOpeningElement">
    >();
    let importsReactThreeFiber = false;
    return {
      Program(node: EsTreeNodeOfType<"Program">) {
        importsReactThreeFiber = hasR3fRuntimeImport(node, context.scopes);
      },
      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
        const elementType = isNodeOfType(node.name, "JSXIdentifier") ? node.name.name : null;
        if (
          !importsReactThreeFiber ||
          !isR3fHostIntrinsic(node) ||
          (elementType !== null && R3F_UNSUPPORTED_SHADOW_LIGHT_NAMES.has(elementType)) ||
          node.attributes.some((attribute) => isNodeOfType(attribute, "JSXSpreadAttribute"))
        ) {
          return;
        }
        const usesShadows = [...SHADOW_OBJECT_PROPERTY_NAMES].some((propertyName) => {
          const attribute = getAuthoritativeJsxAttribute(node.attributes, propertyName);
          return attribute ? readStaticJsxBooleanAttribute(attribute) === true : false;
        });
        if (!usesShadows) return;
        let ancestor = node.parent?.parent ?? null;
        while (ancestor) {
          if (
            isNodeOfType(ancestor, "JSXElement") &&
            isR3fCanvas(ancestor.openingElement, context)
          ) {
            shadowUsersByCanvas.set(ancestor.openingElement, node);
            return;
          }
          ancestor = ancestor.parent ?? null;
        }
      },
      "Program:exit"() {
        for (const [canvas, shadowUser] of shadowUsersByCanvas) {
          if (
            canvas.attributes.some((attribute) => isNodeOfType(attribute, "JSXSpreadAttribute")) ||
            getAuthoritativeJsxAttribute(canvas.attributes, "gl") ||
            getAuthoritativeJsxAttribute(canvas.attributes, "onCreated") ||
            (() => {
              const shadowsAttribute = getAuthoritativeJsxAttribute(canvas.attributes, "shadows");
              return shadowsAttribute
                ? readStaticJsxBooleanAttribute(shadowsAttribute) !== false
                : false;
            })()
          ) {
            continue;
          }
          context.report({
            node: shadowUser,
            message:
              "This object enables castShadow or receiveShadow, but its Canvas leaves shadow maps disabled. Add the Canvas shadows prop",
          });
        }
      },
    };
  },
});

File: native/oxlint/rules/r3f-require-shadows-enabled.rs
use oxc_ast::{AstKind, ast::JSXAttributeItem};
use oxc_diagnostics::OxcDiagnostic;
use oxc_semantic::NodeId;

use crate::{
    context::{ContextHost, LintContext},
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const MESSAGE: &str = "This object enables castShadow or receiveShadow, but its Canvas leaves shadow maps disabled. Add the Canvas shadows prop";
const UNSUPPORTED_SHADOW_LIGHT_NAMES: [&str; 3] =
    ["ambientLight", "hemisphereLight", "rectAreaLight"];

#[derive(Debug, Default, Clone)]
pub struct R3FRequireShadowsEnabled;

impl RuleMeta for R3FRequireShadowsEnabled {
    const NAME: &'static str = "r3f-require-shadows-enabled";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Correctness;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Require Canvas shadow maps for React Three Fiber shadow users.",
    };
}

impl Rule for R3FRequireShadowsEnabled {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx()
    }

    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        if file_is_non_react_jsx_dialect(ctx) || !has_r3f_runtime_import(ctx) {
            return;
        }
        let mut shadow_users_by_canvas = Vec::<(NodeId, oxc_span::Span)>::new();
        for node in ctx.nodes().iter() {
            let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
                continue;
            };
            if !is_r3f_host_intrinsic(opening_element, ctx)
                || matches!(
                    &opening_element.name,
                    oxc_ast::ast::JSXElementName::Identifier(identifier)
                        if UNSUPPORTED_SHADOW_LIGHT_NAMES.contains(&identifier.name.as_str())
                )
                || opening_element
                    .attributes
                    .iter()
                    .any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)))
                || !["castShadow", "receiveShadow"]
                    .iter()
                    .any(|attribute_name| {
                        get_authoritative_jsx_attribute(opening_element, attribute_name, true)
                            .is_some_and(|attribute| {
                                read_static_jsx_boolean_attribute(attribute) == Some(true)
                            })
                    })
            {
                continue;
            }
            let Some(canvas_node_id) = ctx.nodes().ancestors(node.id()).find_map(|ancestor| {
                let AstKind::JSXElement(element) = ancestor.kind() else {
                    return None;
                };
                r3f_canvas_has_public_provenance(&element.opening_element, ctx)
                    .then_some(ancestor.id())
            }) else {
                continue;
            };
            if let Some((_, shadow_user_span)) = shadow_users_by_canvas
                .iter_mut()
                .find(|(candidate_canvas_id, _)| *candidate_canvas_id == canvas_node_id)
            {
                *shadow_user_span = opening_element.span;
            } else {
                shadow_users_by_canvas.push((canvas_node_id, opening_element.span));
            }
        }
        for (canvas_node_id, shadow_user_span) in shadow_users_by_canvas {
            let AstKind::JSXElement(canvas) = ctx.nodes().get_node(canvas_node_id).kind() else {
                continue;
            };
            let opening_element = &canvas.opening_element;
            if opening_element
                .attributes
                .iter()
                .any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)))
                || get_authoritative_jsx_attribute(opening_element, "gl", true).is_some()
                || get_authoritative_jsx_attribute(opening_element, "onCreated", true).is_some()
                || get_authoritative_jsx_attribute(opening_element, "shadows", true).is_some_and(
                    |attribute| read_static_jsx_boolean_attribute(attribute) != Some(false),
                )
            {
                continue;
            }
            ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(shadow_user_span));
        }
    }
}
